### PR TITLE
Fix min max call on array of union types

### PR DESCRIPTION
**Pull request by VincentLanglet**
Closes https://github.com/phpstan/phpstan/issues/8088

The dynamic extensions was considering
```
array<0|1|2|3|4|5|6|7|8|9>
```
and
```
array{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
```
as the same

### CODE REVIEW

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php
Comment (ondrejmirtes) [inline src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: This isn't gonna work if you have union of constant arrays. Please test that scenario + fix it with `TypeUtils::getOldConstantArrays()`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php
Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: I wasn't sure how you wanted that I use getOldConstantArrays.

I handled UnionType in a different way and added a test. Tell me if this is ok this way.

Comment (ondrejmirtes) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: It's not OK, please do not use any `instanceof *Type`. Use `TypeUtils::getOldConstantArrays()` and iterate over the array of types.

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: I tried a solution with getOldConstantArrays now.
But this seems more complicated to me.

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php
Comment (ondrejmirtes) [inline src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: I suspect this if/else is no longer necessary.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php
Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: I just tried, and if I only keep the if part, tests crash, and if I only keep the else part, test fails.

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: Can you explain how you want the code changed @ondrejmirtes ? Because I didn't succeed having my test green as soon as I touch this if/else.

**Comment by VincentLanglet** [comment]
Do you mind taking a new look @ondrejmirtes ? I tried to remove the if part as asked without success, Am I doing something wrong ?

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php
Comment (ondrejmirtes) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: Any code that does `instanceof *Type` is suspicious to me. Just try deleting this piece of code and see what tests fail. And try fixing them in another way.

Recently we've been adding specific methods on Type. It also means that if you rebase on top of 1.9.x, some methods you're currently using will show up as deprecated.

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: Do i need to retarget to 1.9.x ?

Comment (ondrejmirtes) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: Would be ideal, otherwise the build will fail after the merge.

Comment (herndlm) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: `TypeUtils::getOldConstantArrays()` is the one that got deprecated, it would be better to use `Type::getConstantArrays()` on 1.9.x. I want to get rid of usage of any of those, but that's probably not doable everywhere yet :)

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: About the 
```
if ($iterableValueType instanceof UnionType) {
     foreach ($iterableValueType->getTypes() as $innerType) {
          $argumentTypes[] = $innerType;
     }
} else {
     $argumentTypes[] = $iterableValueType;
}
```
code

If I only keep the if part, I get an error, because `getTypes` method only exists in UnionType.
If I only keep the else part, the following test is failing
```
assertType('9', max([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]));
```
because, we're adding to the arguments the iterableValueType `0|1|2|...|8|9` instead of adding all the value types `0`, `1`, ..., `8`, `9` separately.

Comment (herndlm) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: Only slightly related note: I'm still thinking about easier Union type iteration and creating something like a Type::getUnionedTypes() is still in the back of my head. works mostly also as flattenTypes() successor, but it is hiiighly specific

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: > Only slightly related note: I'm still thinking about easier Union type iteration and creating something like a Type::getUnionedTypes() is still in the back of my head. works mostly also as flattenTypes() successor, but it is hiiighly specific

This would be a great solution of what is asked to me (not using `instanceof UnionType` anymore).
Currently, I don't see how I avoid this instanceof check. But maybe you have an idea how I should rewrite the code atm ?

Comment (herndlm) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: not sure, the thing I mentioned can still be done later, I'm still not sure about it. I also don't like the name yet. because sometimes I explicitly want to iterate only union types and sometimes I want to iterate whatever this type is and "spread" the union types, basically e.g. a single IntegerType would be returned as either `[]` or `[IntegerType]` 🤔
sorry, I don't want to drift off here too much, I'll be silent again :)

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: For the record @ondrejmirtes, the same check exists here: https://github.com/phpstan/phpstan-src/pull/1795/files#diff-b1a10b0e3c0ccbea063bae8da98c7f26d862672ddc4e2992db23ba79c5b07121R84-R90
this seems legit to me

Comment (ondrejmirtes) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: No, flattenTypes is 💩 , more likely we need a specific method for min/max, so that IntegerRangeType can also answer correctly.

Comment (herndlm) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: that, of course, would be the best and cleanest solution!

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: > No, flattenTypes is 💩 , more likely we need a specific method for min/max, so that IntegerRangeType can also answer correctly.

So I need to add a new method in the Type interface instead of doing
```
if ($iterableValueType instanceof UnionType) {
    foreach ($iterableValueType->getTypes() as $innerType) {
        $argumentTypes[] = $innerType;
    }
} else {
    $argumentTypes[] = $iterableValueType;
}
```
?

This doesn't sound like an easy things to do. Shouldn't be done in another PR in order to not delay the bugfix since the `instanceof UnionType` code was already here ?

Comment (ondrejmirtes) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: "The window was already broken, I don't need to fix it now."

Sorry, this isn't how I imagine and manage quality software development. The fix needs to make sense and is easier to do once the knowledge of the specific area is already in your headspace.

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: > Sorry, this isn't how I imagine and manage quality software development. The fix needs to make sense and is easier to do once the knowledge of the specific area is already in your headspace.

I'll be more than happy to find another way to write this code but I'll need more help then. Because I didn't understood everything during this discussion. You said "Recently we've been adding specific methods on Type" and "we need a specific method for min/max" so do you want me to add a new method to types ? What should it does ?

Comment (ondrejmirtes) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: @herndlm can help answering that but usually you check recent merged PR history and also the above discussion to get the idea.

Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: I saw Pr like https://github.com/phpstan/phpstan-src/pull/1869 which introduced a new method.

My main concern was about the method to introduce.
I thought about introducing a `getTypes` method to all the Type, in order to write:
```
foreach ($iterableValueType->getTypes() as $innerType) {
    $argumentTypes[] = $innerType;
}
```
But i'm not sure you will like this solution.

You said "we need a specific method for min/max" but at first sight that would mean reproducing/duplicating part of the logic from MinMaxFunctionReturnTypeExtension::processType/compareTypes in every types.

Comment (herndlm) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: sorry, I didn't fully dive into this yet, but it sounds like you can add a `Type::max(Types ...$types): Type` and min methods. moving it to type often could mean slight duplications, but the core logic for arrays should then be split into `ArrayType` and `ConstantArrayType` and we automatically add support for unions and e.g. don't need `Type::getConstantArrays()` & co anymore. The other types are mostly covered via the array traits and we need to think explicitly how the array accessory types like `HasOffsetType` should behave. it is possible then to simply add code for them as well via the new methods.

I didn't think through yet how you can solve the iterableValueType problem that seems to be here. because it sounds like it would just be moved somewhere else. but you have then `max` and `min` methods on type, so maybe that helps somehow. since max and min don't only handle arrays, it means most likely that other types need to implement too somehow, I have no clue yet how, so I'm not sure if that helps much here. Maybe you also need `minArray()` and `maxArray()` methods on type and want to keep the rest here. again, I didn't dive into this, sorry

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php
Comment (ondrejmirtes) [inline src/Type/Php/MinMaxFunctionReturnTypeExtension.php:119]: Finally! This is what I meant that you don't need UnionType here anymore. I'm happy to merge this if the build is green.

**Review by ondrejmirtes — APPROVED** [review]

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php
Comment (VincentLanglet) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php]: > how you can solve the iterableValueType problem

The previous logic was
- If it's an Array
  - if the IterableValueType is an UnionType, iterate over the types.
  - else take the iterableValueType like this.

The logic I neeeded to implements in the ReturnTypeExtension is
- If it's a ConstantArray
  - if the IterableValueType is an UnionType, iterate over the types.
  - else take the iterableValueType like this.
 - If it's just an Array
   -  take the iterableValueType like this.

I just discovered in case it's a ConstantArray, it could just be replaced by the `getValueTypes` method !

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**
